### PR TITLE
cmd: Add bad ssts command for tikv-ctl

### PR DESCRIPTION
**Pull request by Connor1996**
### What problem does this PR solve?
Check corrupted ssts and collect related meta information automatically

### What is changed and how it works?

Use the integrated sst and ldb tool to check the ssts. And use pd_client to get overlapped regions info.

should merge these first
- [x] add sst_dump tools into release https://github.com/tikv/rust-rocksdb/pull/470
- [x] add unsafe remove sst file command for ldb https://github.com/facebook/rocksdb/pull/6678
- [ ] cherry pick unsafe remove sst file command https://github.com/facebook/rocksdb/pull/7335
- [ ] add sst meta dump command for ldb https://github.com/tikv/rocksdb/pull/164

### Related changes
- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List 
- Manual test (add detailed scripts or steps below)

```
$ ./tikv-ctl-new bad-ssts --db=/data3/zbk/kv4/deploy/data/db --pd 172.16.5.39:40601
--------------------------------------------------------
corruption info:
/data3/zbk/kv4/deploy/data/db/000043.sst: Corruption: Bad table magic number: expected 9863518390377041911, found 748006974761695599 in /data3/zbk/kv4/deploy/data/db/000043.sst

sst meta:
43:154204429[2957499 .. 2957499]['7A31' seq:2957499, type:1 .. '7A311B40EE44139D0A9356DE99B95CF51FD2F5B27E08031F9ECA2C3E1553911FCA' seq:2957499, type:1] at level 0 for Column family "write"  (ID 2)

overlap region:
RegionInfo { region: id: 33 end_key: 3110B3C584387715C91ECE7564EAFCE6AAD2FB6C30628FB3F3F6BE50DE6DBBB9 region_epoch { conf_ver: 39 version: 3 } peers { id: 34 store_id: 1 } peers { id: 35 store_id: 6 } peers { id: 36 store_id: 4 } peers { id: 37 store_id: 5 is_learner: true }, leader: None }
RegionInfo { region: id: 13 start_key: 3110B3C584387715C91ECE7564EAFCE6AAD2FB6C30628FB3F3F6BE50DE6DBBB9 end_key: 311B40EFCC2CB4C5911ECF3937D728DED26AE53FA5E61BE04F23F2BE54EACC73 region_epoch { conf_ver: 44 version: 3 } peers { id: 29 store_id: 4 } peers { id: 31 store_id: 5 } peers { id: 38 store_id: 6 }, leader: None }
--------------------------------------------------------
corruption info:
/data3/zbk/kv4/deploy/data/db/000063.sst: Corruption: An external sst file with version 2 have global seqno property with value

sst meta:
63:132906243[3555338 .. 3555338]['7A311B40EFCC2CB4C5911ECF3937D728DED26AE53FA5E61BE04F23F2BE54EACC73' seq:3555338, type:1 .. '7A313030302E25CD5F57252E' seq:3555338, type:1] at level 0 for Column family "write"  (ID 2)

overlap region:
RegionInfo { region: id: 2 start_key: 311B40EFCC2CB4C5911ECF3937D728DED26AE53FA5E61BE04F23F2BE54EACC73 region_epoch { conf_ver: 27 version: 2 } peers { id: 28 store_id: 6 } peers { id: 30 store_id: 1 } peers { id: 32 store_id: 4 } peers { id: 39 store_id: 5 is_learner: true }, leader: None }
--------------------------------------------------------
corruption info:
skip bad line format: @6, while largest seqno in the file is 0
--------------------------------------------------------
finish print
```

### Release note 
- Add bad ssts command for tikv-ctl

### CODE REVIEW

File: cmd/src/bin/tikv-ctl.rs [hunk 1]
```diff
@@ -2267,6 +2317,170 @@ fn run_ldb_command(cmd: &ArgMatches<'_>, cfg: &TiKvConfig) {
...
+            args1.push(format!("--manifest={}", manifest_path));
+        }
+       
+        let mut stdout = BufferRedirect::stdout().unwrap();
+        let mut stderr = BufferRedirect::stderr().unwrap();
+        match run_and_wait_child_process(|| engine::rocks::run_ldb_tool(&args1, &opts)).unwrap() {
+            0 => {},
+            status => {
+                let mut err = String::new();
+                stderr.read_to_string(&mut err).unwrap();
```
Comment (yiwu-arbug) [inline cmd/src/bin/tikv-ctl.rs:2388]: seems we fork ldb just to redirect stderr to stdout? Is there a better way?

**Review by yiwu-arbug — COMMENTED** [review]
Look good to me overall.

**Review by yiwu-arbug — APPROVED** [review]

File: cmd/src/bin/tikv-ctl.rs [hunk 2]
```diff
@@ -1730,24 +1736,69 @@ fn main() {
...
     let mgr = new_security_mgr(&matches);
+    cmd::setup::initial_logger(&cfg);
 
     // Bypass the ldb command to RocksDB.
     if let Some(cmd) = matches.subcommand_matches("ldb") {
         run_ldb_command(&cmd, &cfg);
         return;
     }
 
+    // Bypass the sst dump command to RocksDB.
```
Comment (hunterlxt) [inline cmd/src/bin/tikv-ctl.rs:1787]: Did you link sst_dump into rust-rocksdb?

File: cmd/src/bin/tikv-ctl.rs [hunk 3]
```diff
@@ -2267,6 +2317,170 @@ fn run_ldb_command(cmd: &ArgMatches<'_>, cfg: &TiKvConfig) {
...
+    drop(stderr);
+    drop(stdout);
+
+    for line in corruptions.lines() {
+        v1!("--------------------------------------------------------");
+        // The corruption format may like this:
+        // /path/to/db/057155.sst is corrupted: Corruption: block checksum mismatch: expected 3754995957, got 708533950  in /path/to/db/057155.sst offset 3126049 size 22724
+        v1!("corruption info:\n{}", line);
+        let parts = line.splitn(2, ':').collect::<Vec<_>>();
+        let path = Path::new(parts[0]);
```
Comment (hunterlxt) [inline cmd/src/bin/tikv-ctl.rs:2362]: seems that `parts[0]` will get "/path/to/db/057155.sst is corrupted"?

File: cmd/src/bin/tikv-ctl.rs [hunk 4]
```diff
@@ -2267,6 +2317,170 @@ fn run_ldb_command(cmd: &ArgMatches<'_>, cfg: &TiKvConfig) {
...
+        match path.extension() {
+            Some(ext) if ext.to_str().unwrap() == "sst" => {},
+            _ => { 
+                v1!("skip bad line format: {}", line);
+                continue;
+            },
+        }
+        let sst_file_number = path.file_stem().unwrap().to_str().unwrap();
+        let mut args1 = vec![
```
Comment (hunterlxt) [inline cmd/src/bin/tikv-ctl.rs:2371]: why not use `args`?

**Review by hunterlxt — COMMENTED** [review]
LGTM mostly, I'ill check again tomorrow morning.

**Pull request closed without merging**